server.go

```diff
@@ -561,11 +623,30 @@ func (srv *Server) serveUDP(l *net.UDPConn) error {
 		if len(m) < headerSize {
 			continue
 		}
-		srv.spawnWorker(&response{msg: m, tsigSecret: srv.TsigSecret, udp: l, udpSession: s})
+		wg.Add(1)
+		srv.spawnWorker(&response{
+			msg:        m,
+			tsigSecret: srv.TsigSecret,
+			udp:        l,
+			udpSession: s,
+			wg:         &wg,
+		})
 	}
+
+	return nil
 }
 
 func (srv *Server) serve(w *response) {
+	defer func() {
+		if w.tcp != nil {
+			srv.lock.Lock()
```

Comment:
We now have created a server wide lock that will, sadly, see a lot of contention. Also `defer` is still considered slow

Reply:
I'm wonder why the waitGroup is not enough, why another map to keep track of things?

And as alluded, some benchmarking would help - I'm afraid this slows things down a bit.

Reply:
@miekg Unfortunately the map is necessary to be able to interrupt the `ReadTCP` call below. Without it, everything will just block until the call times out.

Reply:
@miekg While there could be lock contention, I actually don't think it will be significant or an issue. This only occurs in the TCP code path and has no impact on UDP, it will also only be hit once per TCP connection, not per-request.

I could replace the map with a lock-less datastructure, like a linked list perhaps, but that would add complexity; and I don't think `sync.Map` would be appropriate here.

Reply:
maybe drop the defer

Reply:
I merged the defer into the existing defer for the TCP path and removed it for the UDP path.